Instrument pending timers in tests

jamesderlin opened the pull request:
## Description

Flutter widget tests assert if a test completes with timers still
pending.  However, it can be hard to diagnose where a pending timer
came from.  For example, a widget might consume a third-party library
that internally uses a timer.

I added a FakeAsync.pendingTimersDebugInfo getter to quiver
(https://github.com/google/quiver-dart/pull/500).  Make flutter_test
use it.

Additionally modify Flutter's debugPrintStack to take an optional
StackTrace argument instead of always printing StackTrace.current.

## Related Issues

Fixes #4237.

## Tests

Since this adjusts error reporting for specific types of test failures, I did not add any tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [X] No, this is *not* a breaking change.

jamesderlin committed:
Instrument pending timers in tests

Flutter widget tests assert if a test completes with timers still
pending.  However, it can be hard to diagnose where a pending timer
came from.  For example, a widget might consume a third-party library
that internally uses a timer.

I added a FakeAsync.pendingTimersDebugInfo getter to quiver
(https://github.com/google/quiver-dart/pull/500).  Make flutter_test
use it.

Additionally modify Flutter's debugPrintStack to take an optional
StackTrace argument instead of always printing StackTrace.current.

Fixes #4237.

dnfield commented on packages/flutter/lib/src/foundation/assertions.dart:
```diff
@@ -737,19 +737,20 @@ class FlutterError extends Error with DiagnosticableTreeMixin implements Asserti
 ///
 /// The `maxFrames` argument can be given to limit the stack to the given number
-/// of lines. By default, all non-filtered stack lines are shown.
+/// of non-filtered lines. By default, all non-filtered stack lines are shown.
```
nit: I'm not sure it helps to add `non-filtered` in here - I think I'd probably rather remove it from the other part of the sentence, since we don't really have any options for filtering on this method anyway.

jamesderlin replied:
Without it, I think it's unclear how filtering and `maxFrames` interact with each other.  Is filtering done first, and then `maxFrames` frames are printed?  Or is the `StackTrace` truncated first and then filtered?

jamesderlin replied:
I reworded the comment hopefully to both our liking.

dnfield commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -1044,14 +1044,32 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
-      _currentFakeAsync.nonPeriodicTimerCount == 0,
-      'A Timer is still pending even after the widget tree was disposed.'
-    );
+    try {
```
I think the idiom more commonly used here would be something like:

```dart
assert(() {
  if (_currentFakeAsync.periodTimerCount != 0) {
    // Print error/message(s)
    return false;
  }
  // ...
  return true;
}());
```

jamesderlin replied:
I was trying to preserve the existing `AssertionError`s and their messages.

I suppose now that the debugging information should report whether the pending timers are periodic or non-periodic (albeit in a less friendly form), that distinction is less important.

dnfield reviewed (commented):
Couple minor things.

jamesderlin committed:
Update with review feedback from dnfield

* Adjust the wording of the debugPrintStack comment.
* Don't bother trying to preserve the original AssertionError
  messages.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -1044,14 +1044,29 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
+        debugPrint('');
+      }
+      return false;
+    }(),'A Timer is still pending even after the widget tree was disposed.');
```
trivialiest of nits: space missing after comma

jamesderlin replied:
Oops.  Fixed.

> Can we add a test for this? In the past I've found this kind of code ends up being very brittle and because it only gets hit when things fail it takes a long time for anyone to notice it broke.

Okay, I'll see if I can make a relevant test.

jamesderlin replied:
I remember now why I didn't add a test: I don't know how.  I want to test that a test fails, but `test`/`testWidgets` doesn't allow nesting.  I assume that my test will need to essentially run `flutter test` on a failing test and interpret the ouptut.

Do you know of examples of tests of tests that I can use for reference?

Hixie commented:
This is great.

Can we add a test for this? In the past I've found this kind of code ends up being very brittle and because it only gets hit when things fail it takes a long time for anyone to notice it broke.

dnfield commented on packages/flutter/lib/src/foundation/assertions.dart:
```diff
@@ -737,19 +737,21 @@ class FlutterError extends Error with DiagnosticableTreeMixin implements Asserti
 /// [FlutterError.defaultStackFilter].
 ///
-/// The current stack is obtained using [StackTrace.current].
+/// If `stackTrace` is null, dumps the current stack using [StackTrace.current].
```
One last nit: this isn't quite a sentence now.  Maybe 
````
If the `stackTrace` parameter is null, the [StackTrace.current] is used to obtain the stack.
````

jamesderlin replied:
Well, it's a sentence, but maybe it's not in the preferred style and doesn't read as well. =)

I'll change it to your suggestion.

dnfield commented:
You can write a test that fails because it leaks a timer and check the output (and that the error code is -1).  E.g. https://github.com/flutter/flutter/blob/master/dev/bots/test.dart#L87

jamesderlin committed:
Update with more review feedback from dnfield

jamesderlin committed:
Add a test that checks for pending timer failure

Add a test that checks that pending timers in widget tests cause
the test to fail and to print a stack trace for the Timer.

goderbauer commented:
(PR triage): @dnfield says this is waiting for https://github.com/flutter/flutter/pull/37981.

jamesderlin commented:
@dnfield Thanks!  I've added a test.

dnfield commented on dev/automated_tests/test_smoke_test/pending_timer_fail_test.dart:
```diff
@@ -0,0 +1,17 @@
+
+void failingPendingTimerTest() {
+  testWidgets('flutter_test pending timer - negative', (WidgetTester tester) async {
+    Timer(const Duration(seconds: 1), () {});
```
I'm not sure if this makes a difference but can we make the duration longer, like a few minutes?  Just in case this runs on a slow CI machine and actually takes a second to make it all the way through this code (it seems pretty unlikely, but far less likely if the timer is for a few minutes or hours or something).

jamesderlin replied:
I'll change it to 10 minutes.

(BTW, I *meant* to squash #37981 but picked the wrong choice by accident.  Sorry about that.)

dnfield reviewed (approved):
LGTM with nit about the timer duration.

Also, please make sure to squash this when you land it instead of as a merge commit - our CI gets really bogged down by those.

jamesderlin committed:
Increase the timer duration for the pending Timer test
